<a href="https://colab.research.google.com/github/bakhshiali/Quantum-Machine-Learning/blob/main/Data_Embedding.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Embedding

1) Basic Embedding

In [ ]:
!pip install pennylane

In [ ]:
import pennylane as qml
from pennylane import numpy as np

We call proper embedding template

In [ ]:
dev = qml.device('default.qubit', wires=4)

@qml.qnode(dev)
def circuit(feature_vector):
  qml.BasisEmbedding(features=feature_vector,wires=range(4))
  return qml.state()



We define vector X in order to encode binary $1010$ to quantum state $|1010\rangle$

In [ ]:
X=np.array([1,0,1,0])  # Binary representation of 1010

In [ ]:
print(circuit(X))

[0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j
 1.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j]


2) Angle Embedding

We call proper embedding template in the quantum node below, using rotation='Y' to encode vector [0.83,0.24,0.35]

In [ ]:
dev2 = qml.device('default.qubit', wires=3)

@qml.qnode(dev2)
def circuit2(feature_vector):
    qml.AngleEmbedding(features=feature_vector,wires=range(3),rotation='Y')
    qml.Hadamard(0)
    return qml.probs(wires=range(3))



In [ ]:
Y=[0.83,0.24,0.35]

In [ ]:
print(qml.draw(circuit2, expansion_strategy="device")(Y))

0: ──RY(0.83)──H─┤ ╭Probs
1: ──RY(0.24)────┤ ├Probs
2: ──RY(0.35)────┤ ╰Probs



3) Amplitude Embedding

In [ ]:
dev3 = qml.device('default.qubit', wires=2)

@qml.qnode(dev3)
def circuit(feature_vector):
    qml.AmplitudeEmbedding(features=feature_vector,wires=range(2),normalize=True)
    return qml.expval(qml.PauliZ(0))



In [ ]:
Z=[0.24,0.72,0.06,0]

In [ ]:
circuit(Z)

tensor(0.98757764, requires_grad=True)

We write a code to see if we were successful to encode the mentioned vector into amplitudes

In [ ]:
print(circuit(Z))
Z = [0.24, 0.72, 0.06, 0]
amplitude_state = qml.AmplitudeEmbedding(Z, wires=range(2),normalize=True)
print(amplitude_state)

0.9875776397515524
AmplitudeEmbedding(array([0.31524416+0.j, 0.94573249+0.j, 0.07881104+0.j, 0.        +0.j]), wires=[0, 1])


output: AmplitudeEmbedding(array([0.31524416+0.j, 0.94573249+0.j, 0.07881104+0.j, 0.        +0.j]), wires=[0, 1])
***
Amplitudes:<br>
qubit 00: 0.31524416 + 0.0j<br>
qubit 01: 0.94573249 + 0.0j<br>
qubit 10: 0.07881104 + 0.0j<br>
qubit 11: 0.0 + 0.0j<br>
***
0.0j --> real states --> Z vector is encoded into quantum state amplitudes on qubits 0 & 1

In [ ]:
dev3.state

tensor([0.31524416+0.j, 0.94573249+0.j, 0.07881104+0.j, 0.        +0.j], requires_grad=True)